app/frontend/src/utils/hooks.ts

```diff
@@ -0,0 +1,42 @@
+import {
+  Dispatch,
+  MutableRefObject,
+  SetStateAction,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+
+function useIsMounted() {
+  const isMounted = useRef(false);
+
+  useEffect(() => {
+    isMounted.current = true;
+    return () => {
+      isMounted.current = false;
+    };
+  }, []);
+
+  return isMounted;
+}
+
+function useSafeState<State>(
+  isMounted: MutableRefObject<boolean>,
```

Comment:
Why not put `useIsMounted` in this function? Then it doesn't need this argument.

Reply:
I think this is not done because of repetition.

Reply:
Yeah it's mainly to stop the need for the `isMounted` code to run more than once within a component's lifecycle.